src/qutip_cupy/dense.py

```diff
@@ -93,6 +93,211 @@ def adjoint(self):
     def trace(self):
         return self._cp.trace()
 
+    def __add__(left, right):
+        if not isinstance(left, CuPyDense) or not isinstance(right, CuPyDense):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(cp.add(left._cp, right._cp))
+
+    def __matmul__(left, right):
+        if not isinstance(left, CuPyDense) or not isinstance(right, CuPyDense):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(left._cp @ right._cp)
+
+    def __mul__(left, right):
+        dense, number = (left, right) if isinstance(left, CuPyDense) else (right, left)
+        if not isinstance(number, numbers.Number):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(dense._cp * complex(number))
+
+    def __imul__(self, other):
+
+        self._cp.__imul__(other)
+        return self
+
+    def __truediv__(left, right):
+        dense, number = (left, right) if isinstance(left, CuPyDense) else (right, left)
+        if not isinstance(number, numbers.Number):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(dense._cp.__truediv__(number))
+
+    def __itruediv__(self, other):
+        if not isinstance(other, numbers.Number):
+            return NotImplemented
+        self._cp.__itruediv__(other)
+        return self
+
+    def __neg__(self):
+
+        return CuPyDense._raw_cupy_constructor(self._cp.__neg__())
+
+    def __sub__(left, right):
+        if not isinstance(left, CuPyDense) or not isinstance(right, CuPyDense):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(left._cp - right._cp)
+
+    # def __dealloc__(self):
+    #     if self._deallocate and self.data != NULL:
+    #         PyDataMem_FREE(self.data)
+
+
+# @TOCHECK  emti and empty_like are not dispatched in qutip
+# if we do not use them internally here we should remove them
+def empty(rows, cols, fortran):
+    """
+    Return a new Dense type of the given shape, with the data allocated but
+    uninitialised.
+    """
+    order = "F" if fortran else "C"
+    cparr = cp.empty(shape=(rows, cols), dtype=cp.complex128, order=order)
+    return CuPyDense._raw_cupy_constructor(cparr)
+
+
+def empty_like(other, fortran):
+    """
+    Return a new Dense type of the same shape as the given array.
+    """
+    order = "F" if fortran else "C"
+    cparr = cp.empty_like(other, dtype=cp.complex128, order=order)
+    return CuPyDense._raw_cupy_constructor(cparr)
```

Comment:
I suggest we remove `empty` and `empty_like` from this PR, and then add create an issue in this repo to add them and to add the dispatchers to qutip dev.major.

Reply:
When you say the dispatchers  you refer to  `empty`'s and `empty-like`'s dispatcher right? 

Shouldn't we move all the constructors to a separate PR then?

Reply:
Now we are both confused. :)

In this PR, you register specialisations for `diag`, `identity` and `zeros` -- so those are already used.

You don't register specialisations for `empty` and `emtpy_like` so those are unused (and hence I am suggesting not including them in the qutip-cupy code until they can be used).

Reply:
Ah, I see, you are riight the thing is they are not dispathced in `dev.major` either,  see [here](https://github.com/qutip/qutip/blob/dev.major/qutip/core/data/constant.py). I think they are for internal use, maybe I should ask @jakelishman .

Reply:
Not every function is meant to have a dispatcher - `csr.sort` won't either, because it only makes sense for CSR.  `empty` and `empty_like` might not make sense for CuPy, but for `Dense` and `CSR` they're used heavily in internal Cython code to implement various features such as `copy` and as the allocation base for `add`, and so on.  I don't see any particular use for them in Python-only space, though, so there's no dispatcher.

Reply:
The use within `qutip-cupy` would be to create an empty matrix on the GPU without having to copy it from the CPU, but I think we can live without empty for the moment.

Reply:
It'll never make sense for TensorFlow or any potential symbolic data type, so I wouldn't think it'll ever become a dispatcher. I don't see a reason why you'd ever want the low-level object-creation routines to be abstractly dispatched.  How would you go about adding data in in the correct form if you don't know what type of object you've just created?  That's why the medium-level `diags`, `identity`, `one_element`, etc exist, and those guys do have dispatchers.

The expectation is that if you ever wanted to call the super low-level `csr` functions from Python space, it would be in something that looked a little like:
```python
out = csr.copy_structure(other)
out_scipy = out.as_scipy(full=True)
out_scipy.data[:] = ...
return out
```
where you're returning a concrete type, not an abstract one.